### PR TITLE
Fix oblivious analysis of type parameters

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/4040

### CODE REVIEW

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 1]
```diff
@@ -556,7 +556,34 @@ private static bool UsesOblivious(ISymbol symbol)
...
+                    var typeParameters = ((INamedTypeSymbol)symbol).TypeParameters;
+                    foreach (var typeParameter in typeParameters)
+                    {
+                        if (ObliviousDetector.CheckTypeParameterConstraints(typeParameter))
+                        {
+                            return true;
+                        }
+                    }
+
+                    return false;
```
Comment (Evangelink) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:568]: What about

```suggestion
                    return typeParameters.Any(typeParameter => ObliviousDetector.CheckTypeParameterConstraints(typeParameter));
```

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 2]
```diff
@@ -556,7 +556,34 @@ private static bool UsesOblivious(ISymbol symbol)
...
+                }
+                else if (symbol.Kind == SymbolKind.Method)
+                {
+                    if (ObliviousDetector.Instance.Visit(symbol))
+                    {
+                        return true;
+                    }
+
+                    var typeParameters = ((IMethodSymbol)symbol).TypeParameters;
+                    foreach (var typeParameter in typeParameters)
```
Comment (Evangelink) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:578]: Same suggestion here.

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 3]
```diff
@@ -875,17 +879,37 @@ public override bool VisitPointerType(IPointerTypeSymbol symbol)
...
-                    if (symbol.HasReferenceTypeConstraint() && symbol.ReferenceTypeConstraintNullableAnnotation() == NullableAnnotation.None)
+                    if (symbol.IsReferenceType &&
+                        symbol.NullableAnnotation() == NullableAnnotation.None)
+                    {
+                        // Example:
+                        // I<TReferenceType~>
+                        return true;
+                    }
+
+                    return false;
```
Comment (Evangelink) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:893]:
```suggestion
                    // Example:
                    // I<TReferenceType~>
                    return symbol.IsReferenceType &&
                        symbol.NullableAnnotation() == NullableAnnotation.None;
```

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs [hunk 4]
```diff
@@ -875,17 +879,37 @@ public override bool VisitPointerType(IPointerTypeSymbol symbol)
...
+                public static bool CheckTypeParameterConstraints(ITypeParameterSymbol symbol)
+                {
+                    if (symbol.HasReferenceTypeConstraint() &&
+                        symbol.ReferenceTypeConstraintNullableAnnotation() == NullableAnnotation.None)
                     {
+                        // where T : class~
                         return true;
                     }
 
                     foreach (var constraintType in symbol.ConstraintTypes)
```
Comment (Evangelink) [inline src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:906]: Could also be refactored as a `.Any()` call.

File: src/PublicApiAnalyzers/UnitTests/AnnotatePublicApiAnalyzerTests.cs [hunk 5]
```diff
@@ -60,6 +61,134 @@ private async Task VerifyAdditionalFileFixAsync(string source, string oldShipped
...
+}
+";
+
+            var shippedText = @"";
+            var unshippedText = @"#nullable enable
+C<T>.C() -> void
+C<T>
+";
+
+            await VerifyCSharpAsync(source, shippedText, unshippedText, System.Array.Empty<DiagnosticResult>());
```
Comment (Evangelink) [inline src/PublicApiAnalyzers/UnitTests/AnnotatePublicApiAnalyzerTests.cs:80]: I am curious on the why you are passing the empty array. Is it for clarity? Allocation? Something else?

Comment (jcouv) [reply src/PublicApiAnalyzers/UnitTests/AnnotatePublicApiAnalyzerTests.cs]: Will remove. Thanks

File: src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs (hunk 4 above)
Comment (jcouv) [reply src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs:901]: I prefer the explicit loop style. One reason is it's easier to step through.

**Review by Evangelink — APPROVED** [review]

**Review by mavasani — APPROVED** [review]

**Pull request merged**